### ISSUE TITLE
Chrome extension

### ISSUE THREAD

**Issue by oferb** [hasadna/sensory-interface#25]
Creating a Chrome extension, so that we can automatically make tables on the web accessible.

We would analyse the DOM and find interesting tables, and then offer the user to use our tool to access the data.

Still need to figure out the UX: do we add some html elements to the page such as an <a>, so that the user can follow it to sensoryinterface.com, or better yet, to a chrome extension page?

**Comment by S0AndS0** [comment]
Extensions, where supported, should enable the use of more features within the browser, so I think this is a good thing to research further. Here's some documentation for both Chrome and Firefox browser extensions...


- https://developer.chrome.com/extensions/getstarted

- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Your_first_WebExtension


... one thing worth noting is that domains supposedly can block _"Content scripts"_, so it's probably a good idea to attach a disclaimer stating something about how there may be domains that cannot be parsed because of this. An exhaustive list is likely not required, just a simple notice on install should be fine.

**Comment by oferb** [comment]
Yup. Do you have a link describing how domains can block content scripts?

**Comment by S0AndS0** [comment]
Not yet but I've found more information as well as a list of what domains will block content scripts...


https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts


... plus the above linked page goes into how to interconnect extensions scripts with content scripts, and why it'll be required at all for this project. TLDR, extension scripts don't get to see the DOM without the use of API and context scripts.

**Comment by oferb** [comment]
Just to clarify for myself, extension scripts == background scripts, right?
Yah so we'll need content scripts for sure.
We might also need background scripts and to send messages back and forth.
I've actually written a Chrome extension a couple of months ago, so I can offer guidance here.

**Comment by S0AndS0** [comment]
I couldn't say with 100 certainty, each browser vendor seems to have their own nomenclature and even similar concepts may have different behaviors.

That would be excellent! I've not yet written one, keep meaning to, but have lacked a good use case so have only read various docs on the subject; i.e. I've only a bit knowledge and no wisdom in this domain.

**Comment by oferb** [comment]
This is relevant to #35 and #36 as they could run as part of the Chrome extension.

**Comment by S0AndS0** [comment]
After a bit of research it seems that a Chrome extension will be cleaner to implement via a separate branch. Likely it would be easiest to organize by having the extension code within the `master` branch and web site code moved to a `gh-pages` branch.

**Comment by oferb** [comment]
It's fine to develop Chrome extensions on a separate branch, but we will merge it to `master` afterwards.

**Comment by S0AndS0** [comment]
Merging site files with extension code may cause issues for clients and those maintaining the project, that and such an action would also unnecessarily bloat the installation size for an extension.

**Comment by oferb** [comment]
I don't see how that could cause maintenance issues.

**Comment by S0AndS0** [comment]
Consider what'll be more sustainable to maintain, eg. check how file and directory structure is already getting disorganized. Will adding/merging extension files into the same branch as the web-site help or hinder this project?

Speaking from my own experiences of the roughly 55 repositories that I maintain; those that are organized with multi-branch setups encourage better project hygiene, and clean projects encourage more sustainable growth.

**Comment by oferb** [comment]
I see different branches like that are sort of like subfolders.

You could just as well create a subfolder instead of each branch, and have all in the same branch, no?

**Comment by S0AndS0** [comment]
Sorta but not really, it's more like separate work-spaces; the act of checking out an extension branch can be thought of as similar to clearing one's desk for a different set of tasks. Hint, check how I've re-organized the [`gh-pages`](https://github.com/S0AndS0/sensory-interface/tree/gh-pages) branch such that it's setup specifically for working on web-site source code.


Where it **only** about organization I'd agree that a sub-directory would be simpler, however, for libraries, packages, extensions, etc. it's generally good practice to only archive the necessary files. Because separate branches reduces the size of the downloaded/installed archive for clients, as well as third-party services that host said software.


Separate branches also prevents TypeScript from throwing false errors. For example I've other TypeScript projects that have thrown errors simply because something in `node_modules` was also requiring/importing something of the same name, which is super frustrating because the ignore/exclude configurations don't resolve such errors... such things seem to happen because TypeScript by default treats everything as though it'll be in one scope, so in a _mono-branch_ setup every `const` and `let` in one directory then has to be accounted for **and** not repeated in the other.

**Comment by oferb** [comment]
Ok, I get what you say, but I still prefer this project to work in the current way of one master branch.

**Comment by S0AndS0** [comment]
Then a separate repository for a separate project would be the next sustainable option. And if code is to be reused between projects then reusable code should be split into other repositories for installing via a package manager such as NPM.

Multi-branch or multi-repo setups facilitate maintainability, and quicker on-boarding of new developers, much better than mono-repos.

I'm willing to put in some time into transitioning features into a browser extension, because I believe this will enable better and more secure services to be provided to clients.
